File: crawler/security/bounded_io.py
"""Bounded runtime I/O primitives for TASK-022D-2.

This module reads all fixed limits from the sealed TASK-022D-1
TransportBudget. It performs no socket, HTTP, DNS, or environment proxy I/O.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Protocol, Sequence

from crawler.security.transport_budget import BUDGET, BudgetProfile, budget_for

MAX_READ_CHUNK_BYTES = 65536


class BoundedIOError(ValueError):
    def __init__(self, reason_code: str, message: str = "") -> None:
        super().__init__(message or reason_code)
        self.reason_code = reason_code


def _validate_size(value: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise BoundedIOError("invalid_size_value")
    if value < 0:
        raise BoundedIOError("negative_size")
    return value


def check_request_body_size(value: int) -> None:
    observed = _validate_size(value)
    if observed > BUDGET.request_body_bytes:
        raise BoundedIOError("request_body_too_large")


def check_response_headers_size(value: int) -> None:
    observed = _validate_size(value)
    if observed > BUDGET.response_headers_bytes:
        raise BoundedIOError("response_headers_too_large")


def check_response_body_size(profile: str, value: int) -> None:
    selected = budget_for(profile)
    observed = _validate_size(value)
    if observed > selected.response_body_bytes:
        raise BoundedIOError("response_body_too_large")


def _parse_content_length_token(token: str) -> int:
    stripped = token.strip(" \t")
    if not stripped or any(char not in "0123456789" for char in stripped):
        raise BoundedIOError("invalid_content_length")
    return int(stripped)


def _content_length_values(value) -> list[int]:
    if isinstance(value, bool) or value is None:
        raise BoundedIOError("invalid_content_length")
    if isinstance(value, int):
        return [value]
    if isinstance(value, str):
        return [_parse_content_length_token(part) for part in value.split(",")]
    if isinstance(value, (list, tuple)):
        result: list[int] = []
        for item in value:
            if isinstance(item, bool) or not isinstance(item, (int, str)):
                raise BoundedIOError("invalid_content_length")
            if isinstance(item, str):
                result.extend(_parse_content_length_token(part) for part in item.split(","))
            else:
                result.append(item)
        return result
    raise BoundedIOError("invalid_content_length")


def _validate_content_length_values(values: Sequence[int], profile: str) -> int | None:
    selected = budget_for(profile)
    if not values:
        raise BoundedIOError("invalid_content_length")
    first = values[0]
    if first < 0:
        raise BoundedIOError("invalid_content_length")
    for value in values[1:]:
        if value != first:
            raise BoundedIOError("invalid_content_length")
    if first > selected.response_body_bytes:
        raise BoundedIOError("response_body_too_large")
    return first


def validate_content_length(value, profile: str) -> int | None:
    if value is None:
        return None
    parsed = _content_length_values(value)
    return _validate_content_length_values(parsed, profile)


def validate_content_lengths(values: Sequence, profile: str) -> int | None:
    selected = budget_for(profile)
    if not values:
        raise BoundedIOError("invalid_content_length")
    flattened: list[int] = []
    for raw in values:
        if raw is None:
            continue
        flattened.extend(_content_length_values(raw))
    if not flattened:
        raise BoundedIOError("invalid_content_length")
    first = flattened[0]
    if first < 0:
        raise BoundedIOError("invalid_content_length")
    for value in flattened[1:]:
        if value != first:
            raise BoundedIOError("invalid_content_length")
    if first > selected.response_body_bytes:
        raise BoundedIOError("response_body_too_large")
    return first


class TimedReader(Protocol):
    deadline_capable: bool

    def read(self, max_bytes: int, timeout_ms: int) -> bytes:
        ...


class MonotonicClock(Protocol):
    def monotonic_ms(self) -> int:
        ...


class _RealClock:
    def monotonic_ms(self) -> int:
        return int(time.monotonic() * 1000)


@dataclass(frozen=True, slots=True)
class BoundedReadResult:
    data: bytes
    total_read: int


def _validate_clock_sample(value, previous: int | None) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise BoundedIOError("invalid_clock")
    if value < 0:
        raise BoundedIOError("invalid_clock")
    if previous is not None and value < previous:
        raise BoundedIOError("invalid_clock")
    return value


def _sample_clock(clock: MonotonicClock, previous: int | None) -> int:
    try:
        value = clock.monotonic_ms()
    except Exception as exc:
        raise BoundedIOError("invalid_clock") from exc
    return _validate_clock_sample(value, previous)


def _classify_timeout(
    start: int,
    last_activity: int,
    total_timeout_ms: int,
    read_idle_timeout_ms: int,
    now: int,
) -> None:
    if now - start >= total_timeout_ms:
        raise BoundedIOError("total_timeout_exceeded")
    if now - last_activity >= read_idle_timeout_ms:
        raise BoundedIOError("read_idle_timeout")


def _read_bounded(
    profile: BudgetProfile,
    read_idle_timeout_ms: int,
    reader: TimedReader,
    clock: MonotonicClock,
) -> BoundedReadResult:
    start = _sample_clock(clock, None)
    last_clock = start
    last_activity = start
    data = bytearray()
    total = 0
    while True:
        now = _sample_clock(clock, last_clock)
        last_clock = now
        _classify_timeout(start, last_activity, profile.total_timeout_ms, read_idle_timeout_ms, now)

        total_remaining = profile.total_timeout_ms - (now - start)
        timeout_ms = min(read_idle_timeout_ms, total_remaining)
        allowed = min(MAX_READ_CHUNK_BYTES, profile.response_body_bytes - total + 1)
        if allowed <= 0:
            allowed = 1

        try:
            chunk = reader.read(allowed, timeout_ms)
        except BoundedIOError as exc:
            now2 = _sample_clock(clock, last_clock)
            last_clock = now2
            _classify_timeout(start, last_activity, profile.total_timeout_ms, read_idle_timeout_ms, now2)
            raise
        except Exception as exc:
            now2 = _sample_clock(clock, last_clock)
            last_clock = now2
            _classify_timeout(start, last_activity, profile.total_timeout_ms, read_idle_timeout_ms, now2)
            raise BoundedIOError("read_failed") from exc

        now2 = _sample_clock(clock, last_clock)
        last_clock = now2
        _classify_timeout(start, last_activity, profile.total_timeout_ms, read_idle_timeout_ms, now2)

        if not isinstance(chunk, bytes):
            raise BoundedIOError("read_failed")
        if len(chunk) > allowed:
            raise BoundedIOError("response_body_too_large")
        if not chunk:
            return BoundedReadResult(bytes(data), total)
        total += len(chunk)
        if total > profile.response_body_bytes:
            raise BoundedIOError("response_body_too_large")
        data.extend(chunk)
        last_activity = now2


def read_bounded(
    profile: str,
    reader: TimedReader,
    *,
    clock: MonotonicClock | None = None,
) -> BoundedReadResult:
    selected = budget_for(profile)
    if not getattr(reader, "deadline_capable", False):
        raise BoundedIOError("reader_not_deadline_capable")
    return _read_bounded(
        selected,
        BUDGET.read_idle_timeout_ms,
        reader,
        clock or _RealClock(),
    )

File: go-spider/internal/security/bounded_io.go
package security

import (
	"encoding/json"
	"math/big"
	"strings"
	"time"
)

// BoundedIOError is a stable runtime limit failure.
type BoundedIOError struct {
	ReasonCode string
}

func (e *BoundedIOError) Error() string { return e.ReasonCode }

func boundedIOFail(reason string) error { return &BoundedIOError{ReasonCode: reason} }

// TimedReader is the deadline-aware reader contract. Each call receives the
// maximum bytes and the timeout for that read; empty bytes mean EOF.
type TimedReader interface {
	Read(maxBytes int, timeoutMS int) ([]byte, error)
}

const MaxReadChunkBytes = 65536

func checkSizeValue(value int) error {
	if value < 0 {
		return boundedIOFail("negative_size")
	}
	return nil
}

// CheckRequestBodySize rejects request bodies above the sealed budget.
func CheckRequestBodySize(value int) error {
	if err := checkSizeValue(value); err != nil {
		return err
	}
	if value > NewTransportBudget().RequestBodyBytes() {
		return boundedIOFail("request_body_too_large")
	}
	return nil
}

// CheckResponseHeadersSize rejects response headers above the sealed budget.
func CheckResponseHeadersSize(value int) error {
	if err := checkSizeValue(value); err != nil {
		return err
	}
	if value > NewTransportBudget().ResponseHeadersBytes() {
		return boundedIOFail("response_headers_too_large")
	}
	return nil
}

// CheckResponseBodySize rejects a decoded body above the profile budget.
func CheckResponseBodySize(profile BudgetProfile, value int) error {
	if err := checkSizeValue(value); err != nil {
		return err
	}
	if value > profile.ResponseBodyBytes() {
		return boundedIOFail("response_body_too_large")
	}
	return nil
}

func parseContentLengthToken(token string) (*big.Int, error) {
	stripped := strings.Trim(token, " \t")
	if stripped == "" {
		return nil, boundedIOFail("invalid_content_length")
	}
	for _, r := range stripped {
		if r < '0' || r > '9' {
			return nil, boundedIOFail("invalid_content_length")
		}
	}
	value := new(big.Int)
	if _, ok := value.SetString(stripped, 10); !ok {
		return nil, boundedIOFail("invalid_content_length")
	}
	return value, nil
}

func contentLengthValues(value any) ([]*big.Int, error) {
	switch v := value.(type) {
	case nil:
		return nil, nil
	case bool:
		return nil, boundedIOFail("invalid_content_length")
	case int:
		if v < 0 {
			return nil, boundedIOFail("invalid_content_length")
		}
		return []*big.Int{big.NewInt(int64(v))}, nil
	case string:
		parts := strings.Split(v, ",")
		out := make([]*big.Int, 0, len(parts))
		for _, part := range parts {
			parsed, err := parseContentLengthToken(part)
			if err != nil {
				return nil, err
			}
			out = append(out, parsed)
		}
		return out, nil
	case json.Number:
		parts := strings.Split(v.String(), ",")
		out := make([]*big.Int, 0, len(parts))
		for _, part := range parts {
			parsed, err := parseContentLengthToken(part)
			if err != nil {
				return nil, err
			}
			out = append(out, parsed)
		}
		return out, nil
	default:
		return nil, boundedIOFail("invalid_content_length")
	}
}

func contentLengthCompare(values []*big.Int) (*big.Int, error) {
	if len(values) == 0 {
		return nil, boundedIOFail("invalid_content_length")
	}
	first := values[0]
	for _, value := range values[1:] {
		if first.Cmp(value) != 0 {
			return nil, boundedIOFail("invalid_content_length")
		}
	}
	return first, nil
}

func contentLengthWithinProfile(value *big.Int, profile BudgetProfile) (int, error) {
	limit := big.NewInt(int64(profile.ResponseBodyBytes()))
	if value.Cmp(limit) > 0 {
		return 0, boundedIOFail("response_body_too_large")
	}
	return int(value.Int64()), nil
}

// ValidateContentLength validates one Content-Length field value.
func ValidateContentLength(value any, profile BudgetProfile) (int, error) {
	if value == nil {
		return 0, nil
	}
	values, err := contentLengthValues(value)
	if err != nil {
		return 0, err
	}
	first, err := contentLengthCompare(values)
	if err != nil {
		return 0, err
	}
	return contentLengthWithinProfile(first, profile)
}

// ValidateContentLengths validates repeated Content-Length field values.
func ValidateContentLengths(values []any, profile BudgetProfile) (int, error) {
	if len(values) == 0 {
		return 0, boundedIOFail("invalid_content_length")
	}
	flattened := make([]*big.Int, 0)
	for _, raw := range values {
		parsed, err := contentLengthValues(raw)
		if err != nil {
			return 0, err
		}
		flattened = append(flattened, parsed...)
	}
	if len(flattened) == 0 {
		return 0, boundedIOFail("invalid_content_length")
	}
	first, err := contentLengthCompare(flattened)
	if err != nil {
		return 0, err
	}
	return contentLengthWithinProfile(first, profile)
}

// BoundedReadResult is returned only after the whole body is within budget.
type BoundedReadResult struct {
	Data      []byte
	TotalRead int
}

// ReadBounded reads a bounded body using a real monotonic clock.
func ReadBounded(profile BudgetProfile, reader TimedReader) (BoundedReadResult, error) {
	start := time.Now()
	return readBoundedWithClock(profile, NewTransportBudget().ReadIdleTimeoutMS(), reader, func() int64 {
		return time.Since(start).Milliseconds()
	})
}

// ReadBoundedWithClock is the injectable-clock variant used by tests.
func ReadBoundedWithClock(profile BudgetProfile, reader TimedReader, clock func() int64) (BoundedReadResult, error) {
	return readBoundedWithClock(profile, NewTransportBudget().ReadIdleTimeoutMS(), reader, clock)
}

func sampleClock(clock func() int64, previous *int64) (int64, error) {
	value := clock()
	if value < 0 || (previous != nil && value < *previous) {
		return 0, boundedIOFail("invalid_clock")
	}
	if previous != nil {
		*previous = value
	}
	return value, nil
}

func classifyTimeout(start, lastActivity, totalTimeoutMS, readIdleMS, now int64) error {
	if now-start >= totalTimeoutMS {
		return boundedIOFail("total_timeout_exceeded")
	}
	if now-lastActivity >= int64(readIdleMS) {
		return boundedIOFail("read_idle_timeout")
	}
	return nil
}

func readBoundedWithClock(profile BudgetProfile, readIdleMS int, reader TimedReader, clock func() int64) (BoundedReadResult, error) {
	if reader == nil {
		return BoundedReadResult{}, boundedIOFail("reader_not_deadline_capable")
	}
	start, err := sampleClock(clock, nil)
	if err != nil {
		return BoundedReadResult{}, err
	}
	lastClock := start
	lastActivity := start
	data := make([]byte, 0, 1024)
	total := 0
	for {
		now, err := sampleClock(clock, &lastClock)
		if err != nil {
			return BoundedReadResult{}, err
		}
		if err := classifyTimeout(start, lastActivity, int64(profile.TotalTimeoutMS()), int64(readIdleMS), now); err != nil {
			return BoundedReadResult{}, err
		}
		totalRemaining := int64(profile.TotalTimeoutMS()) - (now - start)
		timeoutMS := int64(readIdleMS)
		if totalRemaining < timeoutMS {
			timeoutMS = totalRemaining
		}
		allowed := int64(profile.ResponseBodyBytes()) - int64(total) + 1
		if allowed > MaxReadChunkBytes {
			allowed = MaxReadChunkBytes
		}
		if allowed <= 0 {
			allowed = 1
		}
		chunk, readErr := reader.Read(int(allowed), int(timeoutMS))
		now2, clockErr := sampleClock(clock, &lastClock)
		if clockErr != nil {
			return BoundedReadResult{}, clockErr
		}
		if readErr != nil {
			if err := classifyTimeout(start, lastActivity, int64(profile.TotalTimeoutMS()), int64(readIdleMS), now2); err != nil {
				return BoundedReadResult{}, err
			}
			if e, ok := readErr.(*BoundedIOError); ok {
				return BoundedReadResult{}, e
			}
			return BoundedReadResult{}, boundedIOFail("read_failed")
		}
		if err := classifyTimeout(start, lastActivity, int64(profile.TotalTimeoutMS()), int64(readIdleMS), now2); err != nil {
			return BoundedReadResult{}, err
		}
		if len(chunk) > int(allowed) {
			return BoundedReadResult{}, boundedIOFail("response_body_too_large")
		}
		if len(chunk) == 0 {
			return BoundedReadResult{Data: data, TotalRead: total}, nil
		}
		total += len(chunk)
		if total > profile.ResponseBodyBytes() {
			return BoundedReadResult{}, boundedIOFail("response_body_too_large")
		}
		data = append(data, chunk...)
		lastActivity = now2
	}
}
